keras_rs/src/metrics/ranking_metric.py

```diff
@@ -0,0 +1,200 @@
+        """
+        # TODO (abheesht): Should `y_true` be a dict, with `"mask"` as one key
+        # for parity  with pairwise losses?
+
+        # === Convert to tensors, if list ===
+        # TODO (abheesht): Figure out if we need to cast tensors to
+        # `self.dtype`.
+        if isinstance(y_true, list):
+            y_true = ops.convert_to_tensor(y_true)
+        if isinstance(y_pred, list):
+            y_pred = ops.convert_to_tensor(y_pred)
+        # `sample_weight` can be a scalar too.
+        if isinstance(sample_weight, (list, float, int)):
+            sample_weight = ops.convert_to_tensor(sample_weight)
+
+        # === Process `sample_weight` ===
+        if sample_weight is None:
+            sample_weight = ops.cast(1, dtype=y_pred.dtype)
+
+        y_true_shape = ops.shape(y_true)
+        y_true_rank = len(y_true_shape)
+        sample_weight_shape = ops.shape(sample_weight)
+        sample_weight_rank = len(sample_weight_shape)
+
+        # Check `y_true_rank` first. Can be 1 for unbatched inputs, 2 for
+        # batched.
+        check_rank(y_true_rank, allowed_ranks=(1, 2), tensor_name="y_true")
+
+        # Check `sample_weight` rank. Should be between 0 and `y_true_rank`.
+        check_rank(
+            sample_weight_rank,
+            allowed_ranks=tuple(range(y_true_rank + 1)),
+            tensor_name="sample_weight",
+        )
+
+        # If `sample_weight` rank is 1, it should be of shape `(batch_size,)`.
+        # Otherwise, it should be of shape `(batch_size, list_size)`.
+        if sample_weight_rank == 1:
+            check_shapes_compatible(sample_weight_shape, (y_true_shape[0],))
+            # Uprank this, so that we get per-list weights here.
+            sample_weight = ops.expand_dims(sample_weight, axis=1)
+        elif sample_weight_rank == 2:
+            check_shapes_compatible(sample_weight_shape, y_true_shape)
+
+        # Want to make sure `sample_weight` is of the same shape as
+        # `y_true`.
+        sample_weight = ops.multiply(ops.ones_like(y_true), sample_weight)
+
+        # === Process inputs - shape checking, upranking, etc. ===
+        mask = ops.greater(
+            sample_weight, ops.cast(0, dtype=sample_weight.dtype)
+        )
+        y_true, y_pred, mask, batched = process_inputs(
+            y_true=y_true,
+            y_pred=y_pred,
+            mask=mask,
+            check_y_true_rank=False,
+        )
+
+        # Uprank sample_weight if unbatched.
+        if not batched:
+            sample_weight = ops.expand_dims(sample_weight, axis=0)
+
+        # === Update "invalid" `y_true`, `y_pred` entries based on mask ===
+
+        # `y_true`: assign 0 for invalid entries
+        y_true = ops.where(mask, y_true, ops.zeros_like(y_true))
+        # `y_pred`: assign a value slightly smaller than the smallest value
+        # so that it gets sorted last.
+        y_pred = ops.where(
+            mask,
+            y_pred,
+            -1e-6 * ops.ones_like(y_pred)
```

Comment:
Maybe we should use `keras.config.epsilon()` or make it configuration in `__init__`. What do you think?

Reply:
Yeah, we can use `keras.config.epsilon()`, didn't know it exists :P